sdk/core/azure-core/src/main/java/com/azure/core/util/Option.java

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.NoSuchElementException;
+
+/**
+ * The Option type to describe tri-state. Every Option instance is in one of
+ * the three states: a state representing a non-null-value, null-value, or no-value.
+ *
+ * @param <T> The value type.
+ */
+public final class Option<T> {
+    private static final Option<?> UNINITIALIZED = new Option<>();
+    private static final Option<?> EMPTY = new Option<>(null);
+    private final boolean isInitialized;
+    private final T value;
+
+    /**
+     * Returns an {@link Option} with the specified null-value or non-null-value.
+     *
+     * @param <T> The value type.
+     * @param value the value.
+     * @return an {@link Option} with the value present.
+     */
+    public static <T> Option<T> of(T value) {
+        return value == null ? empty() : new Option<>(value);
+    }
+
+    /**
+     * Returns an {@link Option} with null-value.
+     * <p>
+     * {@code Option.empty()} is a syntactic sugar for {@code Option.of(null)}.
+     * </p>
+     * @param <T> The value type.
+     * @return an {@link Option} with a null-value.
+     */
+    public static <T> Option<T> empty() {
+        @SuppressWarnings("unchecked")
+        Option<T> empty = (Option<T>) EMPTY;
+        return empty;
+    }
+
+    /**
+     * Returns an {@link Option} instance with no-value.
+     *
+     * @param <T> Type of the non-existent value.
+     * @return An Option type with no-value.
+     */
+    public static <T> Option<T> uninitialized() {
+        @SuppressWarnings("unchecked")
+        Option<T> uninitialized = (Option<T>) UNINITIALIZED;
+        return uninitialized;
+    }
+
+    /**
+     * Return {@code true} if this instance is initialized with a null-value or non-null-value,
+     * otherwise {@code false}.
+     *
+     * @return {@code true} if there is a value present, otherwise {@code false}
+     */
+    public boolean isInitialized() {
+        return this.isInitialized;
+    }
+
+    /**
+     * Gets the value in the {@link Option}.
+     *
+     * @return The value.
+     * @throws NoSuchElementException thrown if the {@link Option} has no value.
+     */
+    public T getValue() {
+        if (!this.isInitialized) {
+            throw new NoSuchElementException("No value present");
```

Comment:
This should log and throw.

Reply:
@srnagar  I have a question - Is it a hard rule that any exception has to be wrapped in logger? 

The reason for asking is, this method we will be called from serializer and will have a `isInitialized` check before accessing it. 

My question is, for a type with such internal guarded access, do we need a logger? (think of it as value() method in Optional)

If we still require a logger - do we need ONE logger instance per Option instance given there can be many patch Option properties Or should we prefer a static logger?

Reply:
@anuchandy - logging the exception before throwing is just a general guideline to help us investigate issues when customers contact us. In this case, it's debatable whether it adds too much value, so, I am fine with not logging if this doesn't help much in troubleshooting any issues. If we do want to log, then instead of creating a lot of logger instances, we can simply use a static logger.

Reply:
thanks for the response, yep it's a bit hard to pick here - general guideline vs value addition for the specific case. Let's go without logger and add it if we ever use it in the output model.